src/workerd/api/sockets.c++

```diff
@@ -511,4 +513,117 @@ jsg::Ref<Socket> SocketsModule::connect(
     jsg::Lock& js, AnySocketAddress address, jsg::Optional<SocketOptions> options) {
   return connectImpl(js, kj::none, kj::mv(address), kj::mv(options));
 }
+
+kj::Own<kj::AsyncIoStream> Socket::takeConnectionStream(jsg::Lock& js) {
+  writable->detach(js);
+  readable->detach(js);
```

Comment:
Both of these require that there aren't pending reads or writes in flight and will throw if there are. Just want to make sure that's accounted for here. When a socket is in the middle of a starttls upgrade, for instance, we make sure things are drained before proceeding with the upgrade and detaching the original streams. This is to ensure that the streams are in a good state to actually be detached.

This likely also needs to account for whether a starttls operation is underway, the socket is already closed, etc... basically just some additional guards around whether the socket is in a good state overall to have the underlying connection removed.

Reply:
Let's also add a comment in here so folks coming into the code later can have a heads up about the requirements.